rir/src/compiler/analysis/loop_detection.cpp

```diff
@@ -0,0 +1,106 @@
+#include "loop_detection.h"
+#include "../util/visitor.h"
+
+namespace rir {
+namespace pir {
+
+bool LoopDetection::Loop::comesBefore(Instruction* a, Instruction* b) const {
+    if (a->bb() == b->bb()) {
+        return a->id().idx() < b->id().idx();
+    } else {
+        // ordering vector is reversed: larger means precedes
+        return ordering_.at(a->bb()) > ordering_.at(b->bb());
+    }
+}
+
+LoopDetection::LoopDetection(Code* code, bool determineNesting) {
+    CFG cfg(code);
+    DominanceGraph dom(code);
+    // map of header nodes to tail nodes
+    std::unordered_map<BB*, BBList> tailNodes;
+
+    // find back edges, i.e. edges n->h where h dominates n
+    Visitor::run(code->entry, [&](BB* maybeHeader) {
+        for (const auto& n : cfg.immediatePredecessors(maybeHeader)) {
+            if (dom.dominates(maybeHeader, n)) {
+                if (tailNodes.count(maybeHeader)) {
+                    tailNodes[maybeHeader].push_back(n);
+                } else {
+                    tailNodes.emplace(maybeHeader, BBList({n}));
+                }
+            }
+        }
+    });
+
+    // for each loop header, find the nodes in the loop
+    for (auto& l : tailNodes) {
+        const auto& header = l.first;
+        auto& todo = l.second;
+        std::unordered_map<BB*, size_t> ordering;
+        size_t counter = 0;
+        BBSet body = {header}; // header is part of loop body
+
+#ifdef DEBUG_LOOP_DETECTION
+        std::cerr << "\n========== Detected loop ==========";
+        std::cerr << "\n\tHeader: " << header->id;
+        std::cerr << "\n\tTails:";
+        for (const auto& t : todo) {
+            std::cerr << " " << t->id;
+        }
+        std::cerr << "\n\tBody (in order of discovery):";
+#endif
+
+        // depth-first search starting from the loop tail(s)
+        while (!todo.empty()) {
+            BB* cur = todo.back();
+#ifdef DEBUG_LOOP_DETECTION
+            std::cerr << " " << cur->id;
+#endif
+            todo.pop_back();
+            if (!body.count(cur)) {
+                body.insert(cur);
+                ordering[cur] = counter++;
```

Comment:
If we have multiple tails the ordering works because we know our visitor for including the tails is bfs? If that is the case maybe a comment?

Reply:
I think you're right; this might be broken:

![image](https://user-images.githubusercontent.com/547059/57158422-6012e200-6db1-11e9-9416-38920e199867.png)

In blue, node 5 should have been given a number less than 2.

I should have done my DFS/toplogical sort forwards, starting from the header, rather than going backwards from the tails. (See red numbering.)

In any case, I removed it from this PR, and I'll put a TODO in my loop-peeling branch.